rust-runtime/smithy-types/src/retry.rs

```diff
@@ -68,3 +70,151 @@ pub enum RetryKind {
     /// The response associated with this variant should not be retried.
     NotRetryable,
 }
+
+#[non_exhaustive]
+#[derive(Eq, PartialEq, Debug, Clone, Copy)]
+pub enum RetryMode {
+    Standard,
+    Adaptive,
+}
+
+#[derive(Debug)]
+pub struct RetryModeErr(String);
+
+impl Display for RetryModeErr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "error parsing string '{}' as RetryMode, valid options are 'standard' or 'adaptive'",
+            self.0
+        )
+    }
+}
+
+impl std::error::Error for RetryModeErr {}
+
+impl FromStr for RetryMode {
+    type Err = RetryModeErr;
+
+    fn from_str(string: &str) -> Result<Self, Self::Err> {
+        let string = string.trim();
+        // eq_ignore_ascii_case is OK here because the only strings we need to check for are ASCII
+        if string.eq_ignore_ascii_case("standard") {
+            Ok(RetryMode::Standard)
+        } else if string.eq_ignore_ascii_case("adaptive") {
+            Ok(RetryMode::Adaptive)
+        } else {
+            Err(RetryModeErr(string.to_owned()))
+        }
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub struct RetryConfig {
+    mode: RetryMode,
+    max_attempts: u32,
+}
+
+impl RetryConfig {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn with_retry_mode(mut self, retry_mode: RetryMode) -> Self {
+        self.mode = retry_mode;
+        self
+    }
+
+    pub fn with_max_attempts(mut self, max_attempts: u32) -> Self {
+        self.max_attempts = max_attempts;
+        self
+    }
+
+    pub fn mode(&self) -> RetryMode {
+        self.mode
+    }
+
+    pub fn max_attempts(&self) -> u32 {
+        self.max_attempts
+    }
```

Comment:
should we add `pub fn disabled() -> Self { ... }` that is a quick-path to disabling retries without having to think through what attempts means?